src/ApacheOrcDotNet.OptimizedReaderTest.App/TradeDataSource.cs

```diff
@@ -8,120 +8,182 @@
+                        prices[writeStartIndex + writeIndex] = _buffers.Price.Values[idx];
+
+                        writeIndex++;
+                    }
+                }
+
+                // We only keep looking for values if
+                // nothing has been found before.
+                if (writeIndex == 0 && !_valuesFound)
+                    continue;
+
+                break;
             }
+
+            return writeIndex;
         }
 
-        public int ReadBatch(Span<decimal?> times, Span<decimal?> prices, Span<long?> sizes)
+        public void Reset() => ApplyFilters();
+
+        private void ApplyFilters()
         {
-            var numRows = 0;
-            var startTimeTotalSeconds = (decimal)_startTime.TotalSeconds;
-            var endTimeTotalSeconds = (decimal)_endTime.TotalSeconds;
+            _matches.Clear();
+            _valuesFound = false;
 
-            foreach (var (stripeId, rowGroupIndexes) in _filters)
+            var timeFilterValue = FilterValues.CreateFromTime(min: _lookup.StartTime, max: _lookup.EndTime);
+            var stripeIds = _reader.FilterStripes(_matchCandidates.Keys, _columns.Time, timeFilterValue);
+
+            foreach (var stripeId in stripeIds)
             {
-                foreach (var rowEntryIndex in rowGroupIndexes)
+                var rowGroupIndexes = _reader.FilterRowGroups(_matchCandidates[stripeId], stripeId, _columns.Time, timeFilterValue).ToList();
+
+                foreach (var rowGroupIndex in rowGroupIndexes)
                 {
-                    Task.WhenAll(
-                        _reader.LoadDataAsync(stripeId, rowEntryIndex, _sourceColumnBuffer),
-                        _reader.LoadDataAsync(stripeId, rowEntryIndex, _symbolColumnBuffer),
-                        _reader.LoadDataAsync(stripeId, rowEntryIndex, _timeColumnBuffer),
-                        _reader.LoadDataAsync(stripeId, rowEntryIndex, _sizeColumnBuffer),
-                        _reader.LoadDataAsync(stripeId, rowEntryIndex, _priceColumnBuffer)
-                    ).Wait();
-
-                    for (int idx = 0; idx < _reader.NumValuesLoaded; idx++)
-                    {
-                        var source = _sourceColumnBuffer.Values[idx];
-                        var symbol = _symbolColumnBuffer.Values[idx];
-                        var time = _timeColumnBuffer.Values[idx];
-
-                        if (source == _souce && symbol == _symbol && time >= startTimeTotalSeconds && time <= endTimeTotalSeconds)
-                        {
-                            times[numRows] = _timeColumnBuffer.Values[idx];
-                            sizes[numRows] = _sizeColumnBuffer.Values[idx];
-                            prices[numRows] = _priceColumnBuffer.Values[idx];
-                            numRows++;
-                        }
-                    }
+                    _matches.Enqueue((stripeId, rowGroupIndex));
```

Comment:
This is an interesting solution, essentially precomputing all the rowGroups that are going to need to be processed.  Would it make sense to rework `FilterRowGroups` to operate around that concept and store the `Queue` within?  It feels like this concept would be valuable to all uses of OptimizedReader.